cudamapper/src/overlapper_triggered.cu

```diff
@@ -61,131 +203,162 @@ void OverlapperTriggered::get_overlaps(std::vector<Overlap>& fused_overlaps, std
                 (i.target_position_in_read_ < j.target_position_in_read_));
     };
 
-    thrust::device_vector<Anchor> anchors_buf;
+    // sort on device
+    // TODO : currently thrust::sort requires O(2N) auxiliary storage, implement the same functionality using O(N) auxiliary storage
+    thrust::sort(thrust::device, d_anchors.begin(), d_anchors.end(), comp);
 
-    // chunking anchors array to a size that fits in memory
-    // sort the individual chunks and merge the sorted chunks into host array
-    for (std::vector<Anchor>::iterator anchors_iter = anchors.begin();
-         anchors_iter < anchors.end();
-         anchors_iter += max_anchors_per_block)
-    {
+    // temporary workspace buffer on device
+    thrust::device_vector<char> d_temp_buf;
 
-        auto curblock_start = anchors_iter;
-        auto curblock_end   = anchors_iter + max_anchors_per_block;
-        if (curblock_end > anchors.end())
-            curblock_end = anchors.end();
+    // Do run length encode to compute the chains
+    // note - identifies the start and end anchor of the chain without moving the anchors
+    // >>>>>>>>>
 
-        auto n_anchors_curblock = curblock_end - curblock_start;
+    // d_start_anchor[i] contains the starting anchor of chain i
+    thrust::device_vector<Anchor> d_start_anchor(d_anchors.size());
 
-        // move current block to device
-        anchors_buf.resize(n_anchors_curblock);
-        thrust::copy(curblock_start, curblock_end, anchors_buf.begin());
+    // d_chain_length[i] contains the length of chain i
+    thrust::device_vector<int32_t> d_chain_length(n_anchors);
 
-        // sort on device
-        thrust::sort(thrust::device, anchors_buf.begin(), anchors_buf.end(), comp);
+    // total number of chains found
+    thrust::device_vector<int32_t> d_nchains(1);
 
-        // move sorted anchors in current block back to host
-        thrust::copy(anchors_buf.begin(), anchors_buf.end(), curblock_start);
+    void* d_temp_storage      = nullptr;
+    size_t temp_storage_bytes = 0;
+    // calculate storage requirement for run length encoding
+    cub::DeviceRunLengthEncode::Encode(
```

Comment:
Is exactly the same function called on line 238? I haven't worked much with CUB so I might be missing something

Reply:
Yes it is, every cub method is called twice - 1st call computes storage requirement (if any) and the 2nd call performs the actual computation. This approach in cub lets the caller have control over memory allocations.

Reply:
Ok, thanks for the clarification!